apis/observedobjectcollection/v1alpha1/types.go

```diff
@@ -0,0 +1,126 @@
+type ObservedObjectCollection struct {
+	v1.TypeMeta   `json:",inline"`
+	v1.ObjectMeta `json:"metadata,omitempty"`
+	Spec          ObservedObjectCollectionSpec   `json:"spec"`
+	Status        ObservedObjectCollectionStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// ObservedObjectCollectionList contains a list of ObservedObjectCollection
+type ObservedObjectCollectionList struct {
+	v1.TypeMeta `json:",inline"`
+	v1.ListMeta `json:"metadata,omitempty"`
+	Items       []ObservedObjectCollection `json:"items"`
+}
+
+// ObservedObjectCollectionSpec defines the desired state of ObservedObjectCollection
+type ObservedObjectCollectionSpec struct {
+
+	// ObserveObjects declares what criteria object need to fulfil
+	// to become a member of this collection
+	ObserveObjects ObserveObjectCriteria `json:"observeObjects"`
+
+	// ProviderConfigReference specifies how the provider that will be used to
+	// create, observe, update, and delete this managed resource should be
+	// configured.
+	// +kubebuilder:default={"name": "default"}
+	ProviderConfigReference v12.Reference `json:"providerConfigRef,omitempty"`
+
+	// Template when defined is used for creating Object instances
+	// +optional
+	Template *ObservedObjectTemplate `json:"objectTemplate,omitempty"`
+}
+
+// ObserveObjectCriteria declares criteria for an object to be a part of collection
+type ObserveObjectCriteria struct {
+
+	// APIVersion of objects that should be matched by the selector
+	// +kubebuilder:validation:MinLength:=1
+	APIVersion string `json:"apiVersion"`
+
+	// Kind of objects that should be matched by the selector
+	// +kubebuilder:validation:MinLength:=1
+	Kind string `json:"kind"`
+
+	// Namespace where to look for objects.
+	// If omitted, search is performed across all namespaces.
+	// For cluster-scoped objects, omit it.
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
+
+	// Selector defines the criteria for including objects into the collection
+	Selector v1.LabelSelector `json:"selector"`
+}
+
+// ObservedObjectTemplate represents template used when creating observe-only Objects matching the given selector
+type ObservedObjectTemplate struct {
+
+	// Objects metadata
+	Metadata ObservedObjectTemplateMetadata `json:"metadata,omitempty"`
+}
+
+// ObservedObjectTemplateMetadata represents objects metadata
+type ObservedObjectTemplateMetadata struct {
+
+	// Labels of an object
+	Labels map[string]string `json:"labels,omitempty"`
+
+	// Annotations of an object
+	Annotations map[string]string `json:"annotations,omitempty"`
+}
+
+// ObservedObjectCollectionStatus represents the observed state of a ObservedObjectCollection
+type ObservedObjectCollectionStatus struct {
+	v12.ResourceStatus `json:",inline"`
+
+	// MembershipLabel is the label set on each member of this collection
+	// and can be used for fetching them.
+	// +optional
+	MembershipLabel map[string]string `json:"membershipLabel,omitempty"`
```

Comment:
Do we still need this field in the status? AFAIU, it is always deterministic and we don't have to store this in status.

Reply:
see https://github.com/crossplane-contrib/provider-kubernetes/pull/217#discussion_r1549776009

IMHO, instead of relying on an implicit convention, we can make the label discoverable by users.